src/liballoc/btree/node.rs

```diff
@@ -435,7 +472,9 @@ impl<K, V> NodeRef<marker::Owned, K, V, marker::Leaf> {
     > {
         let node = self.node;
         let ret = self.ascend().ok();
-        Global.dealloc(node.as_opaque(), Layout::new::<LeafNode<K, V>>());
+        if !node.as_ref().is_shared_root() {
```

Comment:
See the `Drop` impl on `IntoIter`, it calls `leaf_node.deallocate_and_ascend()` on the root, so it will always try to deallocate the root node. I should move that check into that drop impl though.

Reply:
I see, thanks for the explanation.

Reply:
It might be cleaner to move this into the IntoIter Drop though, even if it takes another public method. It looks cleaner and `map.rs` is already aware of the shared root anyway.